File: tubes_generation/matlab_version/scatteringcalculator.m
function [scatAmpxy,scatAmpyz,scatAmpxz] = scatteringcalculator(output,scatterers,params,randshifts)
%alltime=tic;
%Generate random displacements within the box
boxlen=params.boxlength;
boxrad=boxlen(1)/2;
%Setup q,theta grid
nq=output.q_and_theta_info(1);
ntheta=output.q_and_theta_info(2);
qmin_exponent=output.q_and_theta_info(3);
qmax_exponent=output.q_and_theta_info(4);
qgrid = logspace(qmin_exponent,qmax_exponent,nq)'*ones(1,ntheta);
thetagrid = ones(nq,1)*linspace(0,pi/2,ntheta);
qmag=qgrid;
qmag(:,:,2)=qmag(:,:,1);
dir1grid=cos(thetagrid);
dir2grid=sin(thetagrid);
dir1grid(:,:,2)=-dir1grid(:,:,1);
dir2grid(:,:,2)=dir2grid(:,:,1);
dir1value=reshape(dir1grid,[],1);
dir2value=reshape(dir2grid,[],1);
qmagvalue=reshape(qmag,[],1);
boxformfactor_sphere=3*(sin(qmagvalue.*boxrad)-(qmagvalue.*boxrad).*cos(qmagvalue.*boxrad))./(qmagvalue.*boxrad).^3;

scatAmpxy = 0;
scatAmpyz = 0;
scatAmpxz = 0;

numruns=size(randshifts,1);

for runind=1:numruns
    if numruns>1
        disp(['Calculating scattering for run#' num2str(runind) ' of ' num2str(numruns) ' runs with random shifts.']);
    end
    scat_XYZ=scatterers(:,3:5)-randshifts(runind,:);
    scat_XYZ=mod(scat_XYZ+boxrad,boxlen)-boxrad;
    datamap=(sum(scat_XYZ.^2,2)<boxrad^2);
    scat_XYZ=scat_XYZ(datamap,:);
    num_scat=length(scat_XYZ(:,1));
    scatAmpxy_curr = 0;
    scatAmpyz_curr = 0;
    scatAmpxz_curr = 0;
    if num_scat
        chunksize = 1024;
        nchunks=ceil(num_scat/chunksize);
        nchunks=pow2(ceil(log2(nchunks)));
        chunksize=ceil(num_scat/nchunks);
        disp(['Total number of chunks (for scattering calc) = ' num2str(nchunks)]);
        numpaddedvals=nchunks*chunksize-num_scat;
        lastvalidchunk=nchunks-floor(numpaddedvals/chunksize);
        scat_XYZ=padarray(scat_XYZ,numpaddedvals,'post');
        chunk_scat_XYZ=pagetranspose(reshape(scat_XYZ',3,chunksize,nchunks));
        parfor n=1:nchunks
            innerlooptime=tic;
            subXYZ=chunk_scat_XYZ(:,:,n);
            if n<lastvalidchunk
                currentchunksize=chunksize;
            elseif n == lastvalidchunk
                currentchunksize=num_scat-(lastvalidchunk-1)*chunksize;
                subXYZ=subXYZ(1:currentchunksize,:);
            else
                disp(['Skipping scattering chunk#' num2str(n) '.']);
                continue;
            end
            qposxy=(qmagvalue*ones(1,currentchunksize)).*(dir1value*subXYZ(:,1)'+dir2value*subXYZ(:,2)');
            qposyz=(qmagvalue*ones(1,currentchunksize)).*(dir1value*subXYZ(:,2)'+dir2value*subXYZ(:,3)');
            qposxz=(qmagvalue*ones(1,currentchunksize)).*(dir1value*subXYZ(:,1)'+dir2value*subXYZ(:,3)');
            resultxy=sum(exp(complex(0,-1)*qposxy)-boxformfactor_sphere*ones(1,currentchunksize),2);
            resultyz=sum(exp(complex(0,-1)*qposyz)-boxformfactor_sphere*ones(1,currentchunksize),2);
            resultxz=sum(exp(complex(0,-1)*qposxz)-boxformfactor_sphere*ones(1,currentchunksize),2);
            scatAmpxy_curr=scatAmpxy_curr+resultxy;
            scatAmpyz_curr=scatAmpyz_curr+resultyz;
            scatAmpxz_curr=scatAmpxz_curr+resultxz;
            disp(['Scattering chunk#' num2str(n) ' of ' num2str(nchunks) '. Time elapsed: ' num2str(toc(innerlooptime)) ' seconds.']);
        end
    end
    scatAmpxy=scatAmpxy+scatAmpxy_curr;
    scatAmpyz=scatAmpyz+scatAmpyz_curr;
    scatAmpxz=scatAmpxz+scatAmpxz_curr;
end

File: tubes_generation/python_version/scattering_calculator.py
import numpy as np
import time
from math import ceil, floor, pi, log2
from multiprocessing import Pool, cpu_count
import psutil

def get_optimal_process_count():
    """Calculate optimal number of processes based on system resources."""
    memory = psutil.virtual_memory()
    cpu_count = psutil.cpu_count()
    available_memory_gb = memory.available / (1024**3)
    memory_per_process_gb = 1  # Conservative estimate for scattering calculation
    
    max_processes_memory = int(available_memory_gb / memory_per_process_gb)
    max_processes_cpu = int(cpu_count * 0.75)
    
    return max(1, min(max_processes_memory, max_processes_cpu))

def process_chunk(args):
    """Process a single chunk of scatterers."""
    subXYZ, currentchunksize, dir1value, dir2value, qmagvalue, boxformfactor_sphere = args
    
    # Reshape inputs for vectorized operations
    q = qmagvalue[:, np.newaxis]
    d1 = dir1value[:, np.newaxis]
    d2 = dir2value[:, np.newaxis]
    xyz = subXYZ[:currentchunksize].T  # Transpose for efficient broadcasting
    
    # Compute q positions for all channels using vectorized operations
    qposxy = q * (d1 * xyz[0] + d2 * xyz[1])
    qposyz = q * (d1 * xyz[1] + d2 * xyz[2])
    qposxz = q * (d1 * xyz[0] + d2 * xyz[2])
    
    # Compute results using vectorized operations
    boxff = boxformfactor_sphere[:, np.newaxis]
    resultxy = np.sum(np.exp(-1j * qposxy) - boxff, axis=1)
    resultyz = np.sum(np.exp(-1j * qposyz) - boxff, axis=1)
    resultxz = np.sum(np.exp(-1j * qposxz) - boxff, axis=1)
    
    return resultxy, resultyz, resultxz

def scatteringcalculator(output, scatterers, params, randshifts):
    boxlen = np.array(params['boxlength'])
    boxrad = boxlen[0] / 2.0

    # Set up q,θ grid
    nq = output['q_and_theta_info'][0]
    ntheta = output['q_and_theta_info'][1]
    qmin_exponent = output['q_and_theta_info'][2]
    qmax_exponent = output['q_and_theta_info'][3]
    
    # Create grids matching MATLAB's meshgrid implementation
    q = np.logspace(qmin_exponent, qmax_exponent, nq)
    theta = np.linspace(0, pi/2, ntheta)
    qgrid, thetagrid = np.meshgrid(q, theta, indexing='ij')
    
    # Create direction grids more efficiently
    dir1grid = np.cos(thetagrid)
    dir2grid = np.sin(thetagrid)
    
    # Stack grids for positive and negative directions
    qmag = np.stack([qgrid, qgrid], axis=2)
    dir1grid = np.stack([dir1grid, -dir1grid], axis=2)
    dir2grid = np.stack([dir2grid, dir2grid], axis=2)
    
    # Flatten arrays efficiently
    dir1value = dir1grid.ravel()
    dir2value = dir2grid.ravel()
    qmagvalue = qmag.ravel()
    
    # Compute form factor
    qrad = qmagvalue * boxrad
    boxformfactor_sphere = 3 * (np.sin(qrad) - qrad * np.cos(qrad)) / (qrad**3)
    
    # Initialize accumulators
    scatAmpxy = np.zeros(len(qmagvalue), dtype=np.complex128)
    scatAmpyz = np.zeros(len(qmagvalue), dtype=np.complex128)
    scatAmpxz = np.zeros(len(qmagvalue), dtype=np.complex128)
    
    numruns = randshifts.shape[0]
    
    for runind in range(numruns):
        if numruns > 1:
            print(f"Calculating scattering for run #{runind+1} of {numruns} runs with random shifts.")
        
        # Adjust scatterer coordinates more efficiently
        scat_XYZ = scatterers[:, 2:5] - randshifts[runind]
        scat_XYZ = (scat_XYZ + boxrad) % boxlen - boxrad
        datamap = np.sum(scat_XYZ**2, axis=1) < (boxrad**2)
        scat_XYZ = scat_XYZ[datamap]
        num_scat = len(scat_XYZ)
        
        if num_scat:
            # Determine chunk size matching MATLAB's implementation
            chunksize = 1024
            nchunks = ceil(num_scat / chunksize)
            nchunks = 2 ** ceil(log2(nchunks))
            chunksize = ceil(num_scat / nchunks)
            
            print(f"Total number of chunks (for scattering calc) = {nchunks}")
            
            # Pad and reshape scatterer array more efficiently
            numpaddedvals = nchunks * chunksize - num_scat
            # Make sure lastvalidchunk is within bounds (0 to nchunks-1)
            lastvalidchunk = min(nchunks - 1, nchunks - floor(numpaddedvals / chunksize))
            
            # Reshape into chunks matching MATLAB's implementation
            scat_XYZ_padded = np.pad(scat_XYZ, ((0, numpaddedvals), (0, 0)), 
                                   mode='constant', constant_values=0)
            chunk_scat_XYZ = scat_XYZ_padded.reshape(nchunks, chunksize, 3)
            
            # Get optimal number of processes
            num_processes = get_optimal_process_count()
            print(f"Using {num_processes} processes for scattering calculation")
            
            # Prepare arguments for parallel processing more efficiently
            args_list = [(chunk_scat_XYZ[n], 
                         chunksize if n < lastvalidchunk else num_scat - (lastvalidchunk-1) * chunksize,
                         dir1value, dir2value, qmagvalue, boxformfactor_sphere)
                        for n in range(lastvalidchunk)]
            
            # Process chunks in parallel
            with Pool(processes=num_processes) as pool:
                results = pool.map(process_chunk, args_list)
            
            # Combine results
            for resultxy, resultyz, resultxz in results:
                scatAmpxy += resultxy
                scatAmpyz += resultyz
                scatAmpxz += resultxz
    
    return scatAmpxy, scatAmpyz, scatAmpxz

